File: languages/dotnet/runtime/Revaly.Sdk/Testing/RapMockTransport.cs
using Revaly.Sdk.Transport;

namespace Revaly.Sdk.Testing;

/// <summary>
/// The first-class no-network test double (DX contract §d): plug it into
/// <see cref="RapClientOptions.Transport"/> and script outcomes with the taxonomy-named
/// builders. Simulates every row of the failover-contract §2 table, both reconcile
/// verdicts and the post-P-2 pending state, and supports scripting consecutive outcomes
/// so merchants can unit-test their failover handler with no network. Asserts on every
/// request that the ADR-SDK-005 User-Agent is present. Synthetic data only (ADR-SDK-020).
/// </summary>
public sealed class RapMockTransport : HttpMessageHandler
{
    private readonly List<(HttpMethod Method, string PathPrefix, MockOperation Operation)> _routes = new();
    private readonly List<RecordedRequest> _requests = new();
    private readonly object _gate = new();

    /// <summary>
    /// Verify on every request that the SDK User-Agent product token leads the header
    /// (on by default — it is part of the contract with platform dashboards).
    /// </summary>
    public bool AssertUserAgent { get; init; } = true;

    /// <summary>Every request observed by the mock, in order.</summary>
    public IReadOnlyList<RecordedRequest> Requests
    {
        get
        {
            lock (_gate)
            {
                return _requests.ToArray();
            }
        }
    }

    /// <summary>Scripts <c>POST /payments</c> (charge).</summary>
    public MockOperation Charge() => Stub(HttpMethod.Post, "/payments");

    /// <summary>Scripts <c>POST /payments/authorize</c>.</summary>
    public MockOperation Authorize() => Stub(HttpMethod.Post, "/payments/authorize");

    /// <summary>Scripts the reconcile GET for one merchant transaction id.</summary>
    public MockOperation Reconcile(string merchantTransactionId)
        => Stub(HttpMethod.Get, $"/transactions/merchant/{merchantTransactionId}");

    /// <summary>Scripts an arbitrary (method, path-prefix) route; the longest matching prefix wins.</summary>
    public MockOperation Stub(HttpMethod method, string pathPrefix)
    {
        lock (_gate)
        {
            foreach (var route in _routes)
            {
                if (route.Method == method && string.Equals(route.PathPrefix, pathPrefix, StringComparison.Ordinal))
                {
                    return route.Operation;
                }
            }

            var operation = new MockOperation();
            _routes.Add((method, pathPrefix, operation));
            return operation;
        }
    }

    /// <inheritdoc />
    protected override async Task<HttpResponseMessage> SendAsync(
        HttpRequestMessage request, CancellationToken cancellationToken)
    {
        var path = request.RequestUri?.AbsolutePath ?? string.Empty;
        var body = request.Content is null
            ? null
            : await request.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);

        var headers = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        foreach (var (name, values) in request.Headers)
        {
            // User-Agent is a space-separated product-token list on the wire; other
            // multi-valued headers join with commas per RFC 9110.
            headers[name] = string.Join(
                string.Equals(name, "User-Agent", StringComparison.OrdinalIgnoreCase) ? " " : ", ",
                values);
        }

        MockOperation operation;
        lock (_gate)
        {
            _requests.Add(new RecordedRequest(request.Method.Method, path, headers, body));

            if (AssertUserAgent)
            {
                var lead = request.Headers.UserAgent.FirstOrDefault();
                if (lead?.Product?.Name != RapUserAgent.ProductName)
                {
                    throw new InvalidOperationException(
                        $"The SDK User-Agent product token must lead every request (ADR-SDK-005); saw '{request.Headers.UserAgent}'.");
                }
            }

            var match = _routes
                .Where(r => r.Method == request.Method
                    && path.StartsWith(r.PathPrefix, StringComparison.Ordinal))
                .OrderByDescending(r => r.PathPrefix.Length)
                .FirstOrDefault();

            operation = match.Operation ?? throw new InvalidOperationException(
                $"Unscripted request: {request.Method} {path}. Script it with Stub()/Charge()/Reconcile().");
        }

        return operation.Next();
    }
}

File: languages/php/runtime/src/Testing/RapMockTransport.php
<?php

declare(strict_types=1);

namespace Revaly\Sdk\Testing;

use GuzzleHttp\Promise\Create;
use GuzzleHttp\Promise\PromiseInterface;
use Psr\Http\Message\RequestInterface;
use Revaly\Sdk\Transport\RapUserAgent;

/**
 * The first-class no-network test double (DX contract §d; runtime-tdd §8): a Guzzle
 * handler that scripts every row of the failover-contract §2 table and both §3
 * verdicts, so a merchant can unit-test their failover handler with no network.
 *
 * Pass it as the client's `transport`:
 *
 * ```php
 * $mock = new RapMockTransport();
 * $mock->charge()->returnsNotProcessed503();
 * $client = new RapClient(apiKey: 'sk-synthetic', transport: $mock);
 * ```
 *
 * Every request is recorded ({@see getRequests()}) and asserted to carry the SDK
 * User-Agent (ADR-SDK-005) — the mock sits where the wire would be, below the runtime
 * middlewares, so it sees exactly what the network would.
 */
final class RapMockTransport
{
    /** @var list<array{method: string, prefix: string, operation: MockOperation}> */
    private array $stubs = [];

    /** @var list<RecordedRequest> */
    private array $requests = [];

    private bool $assertUserAgent = true;

    /** Disables the User-Agent presence assertion (for tests of the assertion itself). */
    public function withoutUserAgentAssertion(): self
    {
        $this->assertUserAgent = false;

        return $this;
    }

    /** Every request this transport has served, in order. */
    public function getRequests(): array
    {
        return $this->requests;
    }

    /** Stubs `POST /payments` (charge). */
    public function charge(): MockOperation
    {
        return $this->stub('POST', '/payments');
    }

    /** Stubs `POST /payments/authorize`. */
    public function authorize(): MockOperation
    {
        return $this->stub('POST', '/payments/authorize');
    }

    /** Stubs the reconcile GET for one merchantTransactionId. */
    public function reconcile(string $merchantTransactionId): MockOperation
    {
        return $this->stub('GET', '/transactions/merchant/' . rawurlencode($merchantTransactionId));
    }

    /** Stubs an arbitrary method + path prefix; the longest matching prefix wins. */
    public function stub(string $method, string $pathPrefix): MockOperation
    {
        $operation = new MockOperation();
        $this->stubs[] = [
            'method' => strtoupper($method),
            'prefix' => $pathPrefix,
            'operation' => $operation,
        ];

        return $operation;
    }

    /** The Guzzle handler contract. */
    public function __invoke(RequestInterface $request, array $options): PromiseInterface
    {
        $this->requests[] = RecordedRequest::from($request);

        if ($this->assertUserAgent) {
            $userAgent = $request->getHeaderLine('User-Agent');
            if (!str_starts_with($userAgent, RapUserAgent::PRODUCT_NAME . '/')) {
                throw new \LogicException(sprintf(
                    'mock: request is missing the SDK User-Agent (ADR-SDK-005); got "%s"',
                    $userAgent,
                ));
            }
        }

        $operation = $this->match($request->getMethod(), $request->getUri()->getPath());
        if ($operation === null) {
            throw new \LogicException(sprintf(
                'mock: no stub for %s %s',
                $request->getMethod(),
                $request->getUri()->getPath(),
            ));
        }

        $outcome = $operation->next($request);
        if ($outcome instanceof \Throwable) {
            return Create::rejectionFor($outcome);
        }

        return Create::promiseFor($outcome);
    }

    private function match(string $method, string $path): ?MockOperation
    {
        $best = null;
        $bestLength = -1;
        foreach ($this->stubs as $stub) {
            if ($stub['method'] !== strtoupper($method)) {
                continue;
            }
            if (!str_starts_with($path, $stub['prefix'])) {
                continue;
            }
            if (strlen($stub['prefix']) > $bestLength) {
                $best = $stub['operation'];
                $bestLength = strlen($stub['prefix']);
            }
        }

        return $best;
    }
}
